### ISSUE TITLE
[Feature Request] Incremental package upload

### ISSUE THREAD

**Issue by wrwg** [aptos-labs/aptos-core#6936]
# 🚀 Feature Request

Incremental population of packages on chain should be supported. A set of modules can be provided on package deployment to either upgrade or add new modules. Since module deletion is not not supported anyway today, this is a conservative extension of existing behavior.

## Motivation

The 64k transaction size limit severely restricts builders with larger Move packages. Packages have to be split introducing security issues because friend visibility is not longer there. 

With this features, builders can upload a larger package in multiple steps. They can also later upgrade the package on a more granular level, e.g. on module base, which safes gas.

**Comment by wrwg** [comment]
@alnoki indeed the scenario you are describing would not work with what I'm proposing. You can only chunk on the level of whole modules with the proposal here. Actually, it must be strongly connected components w.r.t. the friend relations, because Move currently does not allow dangling friend references. 

On the other hand, the solution I'm proposing would allow a smart deployer (tool) to automatically compute the delta (that is the modules actually changed) on package upgrade, and could this way drastically reduce gas cost for repeated upgrade.

Thoughts?

**Comment by movekevin** [comment]
Arbitrarily cutting a package's bytecode into several chunks and uploading them on-chain does intuitively sound messy. This can mean that there's no validation that can protect the data from being corrupted, which developers would only realize upon uploading all the chunks and trying to call code::deploy.

@wrwg If I understand correctly, you're suggesting a more semantic breakdown of the packages where devs can upload module one by one for example (as long as they're in the right dependency order)?

**Comment by wrwg** [comment]
> Arbitrarily cutting a package's bytecode into several chunks and uploading them on-chain does intuitively sound messy. 

Perhaps a generic mechanism for staging data would be an interesting feature for the aptos stdlib.

> @wrwg If I understand correctly, you're suggesting a more semantic breakdown of the packages where devs can upload module one by one for example (as long as they're in the right dependency order)?

Yes, though it is not just the dependency order which can go wrong. For example, you may have A and B in the package, now you compile A and B with some new stuff in A, but only update B on chain. Then bytecode verification will fail. All situations I can think of should be protected by the bytecode verifier. 

In general, we can extend the CLI to compute a delta to the package on chain. This computed subset of uploaded modules should never produce an error. In addition, we can add options for chunking.

**Comment by alnoki** [comment]
> > Arbitrarily cutting a package's bytecode into several chunks and uploading them on-chain does intuitively sound messy.
> 
> Perhaps a generic mechanism for staging data would be an interesting feature for the aptos stdlib.

@wrwg I agree, and here it may be necessary for modules that are larger than 64k: for example complex math libraries or package-in-a-modules (some devs may condense entire packages into a single module, in addition to inlined functions, to reduce function call gas as much as possible).

> Arbitrarily cutting a package's bytecode into several chunks and uploading them on-chain does intuitively sound messy. This can mean that there's no validation that can protect the data from being corrupted, which developers would only realize upon uploading all the chunks and trying to call code::deploy

@movekevin I agree that staging a bytes vector on-chain has potentially messy implications, but for the examples I list above I think this solution may be necessary. Perhaps in the case that staged package bytes fail to deploy, due to a failed bytecode verification, the staging area will simply be flushed?

**Comment by e99243506bigplay** [comment]
Is anyone working on it? I really need to upload very big move package to chain.

**Comment by alnoki** [comment]
Tagging #6905 here rather than commenting there, so that issue can remain stale

**Comment by davidiw** [comment]
Considering the implications of linking and security, this seems like a rather complex introduction. Also note that it has been over 4 months since this issue as produced and I suspect much longer since we started receiving feedback about txn size limitations.

In contrast, with sufficient validation chunking is much easier and can be done reasonably quickly:
* Split up into 32K chunks with a SHA hash for validation
* On-chain validate that the 32K matches the SHA hash and store in a distinct smart vector / object-based linked list
* On the last chunk, aggregate all on-chain chunks and then call publish

**Comment by alnoki** [comment]
> Considering the implications of linking and security, this seems like a rather complex introduction. Also note that it has been over 4 months since this issue as produced and I suspect much longer since we started receiving feedback about txn size limitations.
> 
> In contrast, with sufficient validation chunking is much easier and can be done reasonably quickly:
> 
> * Split up into 32K chunks with a SHA hash for validation
> * On-chain validate that the 32K matches the SHA hash and store in a distinct smart vector / object-based linked list
> * On the last chunk, aggregate all on-chain chunks and then call publish

@davidiw why 32K chunks instead of 64K per the transaction size limit? Is the 64K transaction limit going to be changed?

**Comment by davidiw** [comment]
> @davidiw why 32K chunks instead of 64K per the transaction size limit? Is the 64K transaction limit going to be changed?
:facepalm: 

Due to the way this works out, the chunk cannot be 64K since it needs to be shared across the transaction payload. So I chose the next highest power of 2. Nothing to read into it :stuck_out_tongue: 

https://github.com/aptos-labs/aptos-core/blob/main/types/src/transaction/mod.rs#L692

we already include a good chunk of data in the computation of this 64k.

**Comment by alnoki** [comment]
> > @davidiw why 32K chunks instead of 64K per the transaction size limit? Is the 64K transaction limit going to be changed?
> 
> 🤦
> 
> Due to the way this works out, the chunk cannot be 64K since it needs to be shared across the transaction payload. So I chose the next highest power of 2. Nothing to read into it 😛
> 
> https://github.com/aptos-labs/aptos-core/blob/main/types/src/transaction/mod.rs#L692
> 
> we already include a good chunk of data in the computation of this 64k.

@davidiw does it need to be a power of two? It seems somewhat burdensome that a 63k package takes once transaction but a 65k one takes 3

**Comment by davidiw** [comment]
No, I just like powers of 2. We can definitely optimize it. I was also just sketching out the concept.

**Comment by wrwg** [comment]
> Considering the implications of linking and security, this seems like a rather complex introduction. Also note that it has been over 4 months since this issue as produced and I suspect much longer since we started receiving feedback about txn size limitations.
> 
> In contrast, with sufficient validation chunking is much easier and can be done reasonably quickly:
> 
> * Split up into 32K chunks with a SHA hash for validation
> * On-chain validate that the 32K matches the SHA hash and store in a distinct smart vector / object-based linked list
> * On the last chunk, aggregate all on-chain chunks and then call publish

This is pretty hilarious. Please take the discussion about chunking into the other issue @alnoki opened. I do not have cycles to work on this t(not very complex) feature at this point, and it appears your entire larger remaining team also doesn't have. So lets move on.

**Comment by alnoki** [comment]
For reference I believe @wrwg was referring to #6905 in his last message

**Comment by wrwg** [comment]
Reopening, I still think we want to do this eventually
